### ISSUE TITLE
Consider splitting updating vs writing Scatter

### ISSUE THREAD

**Issue by tzerrell** [plaidml/plaidml#1612]
The current Scatter (as of https://github.com/plaidml/plaidml/pull/1610) has a unified API with a parameter controlling whether to override or sum into existing results. For scatters where we know summing isn't needed, we currently need to pass a data tensor for API consistency with summing scatter (rather than for algorithmic reasons). We could instead split scatter into two versions and thereby avoid passing the data tensor to a non-updating scatter. Consider whether we want to make this change.

More details available in the linked PR.

**Comment by haoyouab** [comment]
Another point that is worth noting is the behavior of updating scatter with duplicate indices.

According  to [tf.tensor_scatter_nd_update](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update):

> If indices contains duplicates, then their updates are accumulated (summed).

while [tf.compat.v1.scatter_update](https://www.tensorflow.org/api_docs/python/tf/compat/v1/scatter_update) shows some conflicts:

> If values in ref is to be updated more than once, because there are duplicate entries in indices, the order at which the updates happen for each value is undefined.

The current implementation  does not take summation into account (openvino ops either) in updating modes but I think we should be aware of this in case any changes on these.

**Comment by flaub** [comment]
Is it possible to have a single primitive but then have 2 APIs that use this primitive?

**Comment by tzerrell** [comment]
> Is it possible to have a single primitive but then have 2 APIs that use this primitive?

I think the answer is no, but perhaps MLIR can be flexible in ways I don't know about.

The fundamental distinction between the two cases is that the update case requires us to read an initial `data` tensor of values that will be updated (or used as the default values if never updated). Whereas the write case will create a new tensor  that is everywhere zero (except where scatter writes), so no initial `data` needs to be read.

I don't know how to unify that into a single primitive without forcing a `data` tensor to passed in the write case even though it will never be used. But perhaps there's a mechanism in MLIR for optional inputs that I'm unaware of? In which case we should use that in creating a unified primitive.

**Comment by tzerrell** [comment]
I suppose it might also be possible to pass an empty `data` tensor in the write case (and be sure to get the sizing data from the output shape, _not_ the `data` tensor)? Not sure if that creates too confusing of an interface (i.e. because the shape of the `data` represents the output shape in the update case, but is unrelated to the output shape in the write case).

**Comment by jbruestle** [comment]
Actually, I would argue that in both cases we should pass an initalization value.  Users that want 0 initialization can pass that as initialization, which makes the behavior explicit. However, it's very useful to do 'assignment' scatter onto an existing tensor (to write to just a subset for example).  So in my mind, the right 'generic' way to write scatter is to have:

1) Initialization tensor (which also defines the output size)
2) Indexes + Values for the actual scattering
3) An 'aggregation' type, which can use our the standard atomic reduce operation.

We can replace the write with atomicrmw in all cases.  If we happen to use 'assign' as the atomic op, the lowering can remove the read (as it is redundant).  But that way both styles of scatter are handled the same way, and the behavior is also identical to contraction operations. 

This was of representing scatter seems to handle all the case, and is consistent and simple to implement.  Moreover, I don't see any complexity of getting it's perf to be identical to other versions with simple optimizations.

**Comment by haoyouab** [comment]
> Actually, I would argue that in both cases we should pass an initalization value. Users that want 0 initialization can pass that as initialization, which makes the behavior explicit. However, it's very useful to do 'assignment' scatter onto an existing tensor (to write to just a subset for example). So in my mind, the right 'generic' way to write scatter is to have:
> 
> 1. Initialization tensor (which also defines the output size)
> 2. Indexes + Values for the actual scattering

I agree. Passing in an initialization tensor in both cases indeed makes the behavior consistent and more explicit. I will send a PR with small changes to support this later.

> 3. An 'aggregation' type, which can use our the standard atomic reduce operation.
> 
> We can replace the write with atomicrmw in all cases. If we happen to use 'assign' as the atomic op, the lowering can remove the read (as it is redundant). But that way both styles of scatter are handled the same way, and the behavior is also identical to contraction operations.
> 
> This was of representing scatter seems to handle all the case, and is consistent and simple to implement. Moreover, I don't see any complexity of getting it's perf to be identical to other versions with simple optimizations.


It also makes sense to simplify the code in `tile_to_pxa` conversion by using a single `pxa::PxaReduceOp` and `op.agg()` to determine the exact ops. There shouldn't be any impact on performance as there's no underlying changes but just code structure optimization.

The problems here are in the conversions `TileAccumulate` and `ConvertPXAToAffine`. See snippets below:

#### TileAccumulate:
```
VERBOSE-1 [default] Failed stride info: op = %5 = index_cast %4 : i32 to index
// *** IR Dump After TileAccumulate *** ('func' operation: @main)
--snippet--
    %1 = affine.parallel (%arg5) = (0) to (4) step (4) reduce ("assign") -> (memref<8xf32>) {
      %2 = affine.parallel (%arg6) = (%arg5) to (%arg5 + 4) reduce ("assign") -> (memref<8xf32>) {
        %3 = pxa.load %arg3[%arg6] : memref<4xf32>
        %4 = pxa.load %arg2[%arg6] : memref<4xi32>
        %5 = index_cast %4 : i32 to index
        %6 = pxa.reduce addf %3, %arg4[%5] : memref<8xf32>
        affine.yield %6 : memref<8xf32>
      }
      affine.yield %2 : memref<8xf32>
    }
    return %1 : memref<8xf32>
  }
--snippet--
}

```

#### ConvertPXAToAffine:
```
// *** IR Dump After ConvertPXAToAffine ***
--snippet--
    affine.parallel (%arg5) = (0) to (1) {
      affine.for %arg6 = 0 to 4 {
        %0 = affine.load %arg3[%arg6] : memref<4xf32>
        %1 = affine.load %arg2[%arg6] : memref<4xi32>
        %2 = index_cast %1 : i32 to index
        %3 = affine.load %arg4[%2] : memref<8xf32>
        %4 = addf %3, %0 : f32
        affine.store %4, %arg4[%2] : memref<8xf32>
      }
    } {tags = {cpuThread}}
    return
  }
--snippet--
}

error: 'affine.load' op index must be a dimension or symbol identifier
note: see current operation: %3 = "affine.load"(%arg4, %2) {map = affine_map<(d0) -> (d0)>} : (memref<8xf32>, index) -> f32

```

It seems `affine.load` doesn't accept values read from arbitrary tensors as indexes. I think this is related to [Restrictions on Dimensions and Symbols](https://mlir.llvm.org/docs/Dialects/Affine/#restrictions-on-dimensions-and-symbols) from specs of affine dialect. Any thoughts regarding this?

**Comment by tzerrell** [comment]
I think this may need the attention of someone with more MLIR Affine dialect experience than me (@jbruestle perhaps?), but I see `index_cast` to `store` to `affine.yield` in working code; perhaps there's something analogous to the load case? That said, I'm confused as to how even that setup works, as it seems like indexing based on a value read from a tensor is outside the scope of Affine operations, so I suspect I'm overlooking something fundamental.

**Comment by flaub** [comment]
It does seem like the `%2` value passed as an index to `affine.load` is illegal, because there's not way to guarantee that it is in fact a linear/affine load. Could we try using a `std.load` here? This is probably just a bug in the `tile.index` lowering to PXA. Although I'm curious why we haven't seen this before.

**Comment by flaub** [comment]
Ya, so a bit more thinking on this, it seems like it's not legal to turn a `tile.index` op into a standard `affine.parallel` loop with a `pxa.reduce` for storage. We probably need to have PXA adhere to the same affine dialect restrictions on dimensions and symbols.

**Comment by haoyouab** [comment]
Yes. Just replacing `affine.[load|store]` with `std.[load|store]` works fine in this case (`GatherOp` and `ScatterOp` are doing so now). See snippet below:

```
// *** IR Dump After ConvertPXAToAffine ***
--snippet--
    affine.parallel (%arg5) = (0) to (1) {
      affine.for %arg6 = 0 to 4 {
        %0 = affine.load %arg3[%arg6] : memref<4xf32>
        %1 = affine.load %arg2[%arg6] : memref<4xi32>
        %2 = index_cast %1 : i32 to index
        %3 = load %arg4[%2] : memref<8xf32>
        %4 = addf %0, %3 : f32
        store %4, %arg4[%2] : memref<8xf32>
      }
    } {tags = {cpuThread}}
    return
  }
--snippet--
}
```
So I wonder if we should keep the current way  that `tile_to_pxa` deals with `IndexOp` for now @jbruestle @flaub ?

**Comment by jbruestle** [comment]
Yes, you must use a non-affine memory access since it is *not* an affine access since it's indirected and can't be known until runtime.  The code above is correct since the update location only depends on the serial loop index, but there is probably some argument that we should use atomicrmw rather than load/add/store if we ever want to have an affine.parallel loop where multiple parallel threads may potentially add into the same location.  But other than that, looks great.
